src/OrchardCore.Modules/OrchardCore.Users/Migrations.cs

```diff
@@ -27,86 +27,90 @@ public int Create()
+                "DocumentId",
+                "NormalizedUserName",
+                "NormalizedEmail",
+                "IsEnabled",
+                "UserId")
             );
 
-            SchemaBuilder.AlterIndexTable<UserIndex>(table => table
-                .CreateIndex("IDX_UserIndex_UserId", "DocumentId", "UserId")
+            SchemaBuilder.CreateReduceIndexTable<UserByRoleNameIndex>(table => table
+               .Column<string>("RoleName")
+               .Column<int>("Count")
             );
 
-            SchemaBuilder.AlterIndexTable<UserIndex>(table => table
-                // This index will be used for lookups when logging in.
-                .CreateIndex("IDX_UserIndex_UserName", "DocumentId", "NormalizedUserName")
-            );
+            SchemaBuilder.CreateMapIndexTable<UserByLoginInfoIndex>(table => table
+                .Column<string>("LoginProvider")
+                .Column<string>("ProviderKey"));
 
-            SchemaBuilder.AlterIndexTable<UserIndex>(table => table
-                .CreateIndex("IDX_UserIndex_Email", "DocumentId", "NormalizedEmail")
+            SchemaBuilder.AlterIndexTable<UserByLoginInfoIndex>(table => table
+                .CreateIndex("IDX_UserByLoginInfoIndex_DocumentId",
+                "DocumentId",
+                "LoginProvider",
+                "ProviderKey")
             );
 
-            SchemaBuilder.CreateReduceIndexTable<UserByRoleNameIndex>(table => table
-               .Column<string>("RoleName")
-               .Column<int>("Count")
+            SchemaBuilder.CreateMapIndexTable<UserByClaimIndex>(table => table
+               .Column<string>(nameof(UserByClaimIndex.ClaimType))
+               .Column<string>(nameof(UserByClaimIndex.ClaimValue)),
+                null);
+
+            SchemaBuilder.AlterIndexTable<UserByClaimIndex>(table => table
+                .CreateIndex("IDX_UserByClaimIndex_DocumentId",
+                "DocumentId",
+                nameof(UserByClaimIndex.ClaimType),
+                nameof(UserByClaimIndex.ClaimValue))
             );
 
-            return UpdateFrom1();
+            // Shortcut other migration steps on new content definition schemas.
+            return 9;
         }
 
+        // This code can be removed in a later version.
         public int UpdateFrom1()
         {
             SchemaBuilder.CreateMapIndexTable<UserByLoginInfoIndex>(table => table
                 .Column<string>("LoginProvider")
                 .Column<string>("ProviderKey"));
+
             return 2;
         }
 
+        // This code can be removed in a later version.
         public int UpdateFrom2()
         {
             SchemaBuilder.CreateMapIndexTable<UserByClaimIndex>(table => table
                .Column<string>(nameof(UserByClaimIndex.ClaimType))
                .Column<string>(nameof(UserByClaimIndex.ClaimValue)),
                 null);
 
-            // Return 8 here to skip migrations on new database schemas.
-            return 8;
```

Comment:
We did this because in the above `Create()`, we were using `return UpdateFrom1()`, which is not a good pattern.

After adding a migration step, `Create()` should be updated to do all the things, and always shortcut migration steps.

This is what i did above by adding in `Create()` the columns added in `UpdateFrom2()`, and then shortcut the steps.

Reply:
Ah I see, yes it was a bad / weird pattern, so you have removed it. great.